### PR TITLE
Dual GPS Heading support via AP_Periph and UAVCAN lib

### PR DESCRIPTION
**Pull request by bugobliterator**
After the implementation one will be able to use two CAN AP_Periph GPS modules as dual heading CAN and also be able to combine them with Serial Ublox Compasses as Rover or Base

Tests:
- [x]  CAN(base) + CAN(rover) Tested working with 2 HerePro GPSes
- [ ]  CAN(base) + Serial(rover)
- [ ]  Serial(base) + CAN(rover)

How it works:

Following parameters need to be setup in Ardupilot Vehicle:
```
GPS_TYPE = 22 (UAVCAN Base)
GPS_TYPE2 = 23 (UAVCAN Rover)
GPS1_CAN_OVRIDE = <Base NODEID> (need to be one from GPS_CAN_NODEID1 or GPS_CAN_NODEID2)
GPS2_CAN_OVRIDE = <Rover NODEID> (need to be one from GPS_CAN_NODEID1 or GPS_CAN_NODEID2)
GPS_AUTO_SWITCH = 1
GPS_AUTO_CONFIG = 2  (GPS UAVCAN Driver will automatically set required parameters on UAVCAN GPS)
```

Set the GPS_POS1_X/Y/Z and GPS_POS2_X/Y/Z parameters for the GPSs (see Sensor Position Offset are here)

```
AHRS_EKF_TYPE = 3 (to use EKF3)
EK2_ENABLE = 0 (to disable EKF2)
EK3_ENABLE = 1 (to enable EKF3)
EK3_SRC1_YAW = 2 (“GPS”) or 3 (“GPS with Compass Fallback”)
```

On the connections side: One CAN port each from HerePros was connected to CubeOrange CAN1 port. And the remaining ports on HerePros were connected together. This configuration removes the traffic burden on single CAN line. Also note that while using AP_Periph firmware from this branch no configuration is required on HerePros themselves. And also one doesn't need to worry about which CAN port is to be used for connecting to autopilot and which to other GPS, this is automatically detected and setup for user as well at boot. Basically AP_Periph setup is zero touch. All configurations are on Autopilot side.

Ref. https://ardupilot.org/copter/docs/common-gps-for-yaw.html

### CODE REVIEW

**Comment by MallikarjunSE** [comment]
Hey @bugobliterator 
I tried this branch with 2×F412 CAN nodes with F9P.
As soon as I sent GPS_TYPE=22, 23, Mission planner stops showing GPS fix. When set as 9(UAVCAN), I get both GPS fix. 
I can see GPS config on nodes is changed to GPS_TYPE=17, 18.
Is there a way to check whether my setup is right?
Will mission planner output any error messages?
Thank you!

**Comment by bugobliterator** [comment]
> Hey @bugobliterator
> I tried this branch with 2×F412 CAN nodes with F9P.
> As soon as I set GPS_TYPE=22, 23, Mission planner stops showing GPS fix(Goes to no fix). When set as 9(UAVCAN), I get both GPS fix.
> I can see GPS config on nodes is changed to GPS_TYPE=17, 18.
> Is there a way to check whether my setup is right?
> Will mission planner output any error messages?
> Thank you!

Did you set define GPS_MOVING_BASELINE 1 in your hwdef? Also need to update Cube firmware built from this PR.

**Comment by MallikarjunSE** [comment]
Yes. I did. And I built Cube orange firmware from this PR. 
I'll try the updated PR again. Thank you!

**Review by tridge — CHANGES_REQUESTED** [review]
need to update AP_GPS::gps_yaw_deg() to play the same instance trick as used for ublox backend

File: libraries/AP_GPS/AP_GPS.cpp [hunk 1]
```diff
@@ -855,21 +862,29 @@ void AP_GPS::update_instance(uint8_t instance)
...
-#if GPS_MAX_RECEIVERS > 1
+#if GPS_MOVING_BASELINE
     if (drivers[instance] && _type[instance] == GPS_TYPE_UBLOX_RTK_BASE) {
         // see if a moving baseline base has some RTCMv3 data
         // which we need to pass along to the rover
         const uint8_t *rtcm_data;
         uint16_t rtcm_len;
+        bool clear_rtcmv3_buffer = false;
         if (drivers[instance]->get_RTCMV3(rtcm_data, rtcm_len)) {
+            if (_mbl_data_cb != nullptr) {
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS.cpp:873]: i think this is simpler by exposing clear_RTCMv3() on AP_GPS, and then AP_Periph calls gps.get_RTCMv3(instance, rtcm_data, rtcm_len) and then calls the clear

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by tridge — APPROVED** [review]

**Comment by MallikarjunSE** [comment]
Hey @bugobliterator 
Should I have to build firmware for Cube from this PR or Latest master will work fine?
For AP-Periph I have built from latest master. Is that okay?
Thank you!

**Comment by bugobliterator** [comment]
> Hey @bugobliterator 
> Should I have to build firmware for Cube from this PR or Latest master will work fine?
> For AP-Periph I have built from latest master. Is that okay?
> Thank you!
> 

Yes, you need to build firmware for all hardware involved using latest master.

**Comment by MallikarjunSE** [comment]
Thanks. I tried with all hardware on latest master. I couldn't get it to work. EKF3 still waiting for GPS config data messages. I have set CAN order, GPS_TYPE to 22, 23. GPS auto config to 2, using EKF3, EK3_SRC1_YAW is 2.

**Comment by MallikarjunSE** [comment]
I built from your PR. MP shows "no fix" with both GPS detected. I don't get fix even though I get fix in GPS when checked from UAVCAN monitor. But F412 node hangs on AP-Periph built from your PR. 
Any suggestions?
Thank you!

**Comment by MallikarjunSE** [comment]
Hey @bugobliterator 
Should both of these need to be defined? 
define GPS_MOVING_BASELINE 1
define GPS_UBLOX_MOVING_BASELINE TRUE
Also same goes on Cube hwdef?

**Comment by bugobliterator** [comment]
@MallikarjunSE I will double check. I need to find time to prepare a wiki page for it.

**Comment by Hwurzburg** [comment]
@bugobliterator @tridge 
the last comments from a user was that this was not working for him....we need a wiki page update! what was he doing wrong...from my reading he seemed to have things setup correctly...is it a connection issue? 

my reading of the above is that the CAN port on the autopilot needs a splitter to attach to both GPS so that they are on the same physical CAN bus, and that the two GPS must be connected together on their second bus...is this correct?
what, if any, other params not stated above in original comment must be setup?

also, to allow this to be used on an F9P UAVCAN GPS (like taking an MatekL431 and attaching the F9P GPS to it), the latest stable periph firmware must be used and MB_CAN_PORT set to ? under what conditions do you set 1 or 2?

**Pull request merged**